[export] Unify single and multiple return for hops

yiming0416 opened the pull request:
Summary: Introduce `is_hop_single_tensor_return` field to the `Node` class in serialization so that during deserialization when there is a single return, we know whether it is a tuple of a single element or a single element.

Test Plan:
```
buck2 run @mode/dev-nosan sigmoid/inference/test:e2e_test_cpu -- -r E2ETestCPUCond
buck2 run @mode/dev-nosan sigmoid/inference/test:test_passes -- -r test_const_folding2
```

Differential Revision: D66991624




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

avikchaudhuri commented on torch/_export/serde/schema.py:
```diff
@@ -212,6 +212,7 @@ class Node:
     inputs: Annotated[List[NamedArgument], 20]
     outputs: Annotated[List[Argument], 30]
     metadata: Annotated[Dict[str, str], 40]
+    is_single_tensor_return: Annotated[Optional[bool], 50] = None
```
rename to `returns_single_tensor` maybe? sounds better to me...

yiming0416 replied:
we use `is_single_tensor_return` in the dataclass of `Graph`, so I'm keeping it same here.

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -575,11 +576,25 @@ def handle_call_function(self, node: torch.fx.Node):
                 inputs=self.serialize_hoo_inputs(node.args, node.kwargs),
                 outputs=self.serialize_hoo_outputs(node),
                 metadata=self.serialize_metadata(node),
+                is_single_tensor_return=_is_hop_single_tensor_return(node),
```
looks like we only use this for hops?

yiming0416 replied:
yes we only use this for hops

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -1798,15 +1819,20 @@ def deserialize_node(self, serialized_node: Node, target: Callable) -> None:
                 serialized_node.outputs[0].as_tensor.name
                 if len(serialized_node.outputs) == 1
                 and hasattr(serialized_node.outputs[0], "as_tensor")
+                and getattr(serialized_node, "is_single_tensor_return", True)
```
This is weird, and probably further reinforces that this is a HOP-only concept and that should be reflected in the name. Because for every other op the field will be `None` even if the op returns a tuple (can it)?

yiming0416 replied:
Yes you're right. This is a HOP-only concept. But in the `Node` schema we didn't differentiate hop and other nodes.

Probably we should rename this to `is_hop_single_tensor_return`? So that it makes sense for every other op the field will always be `None`

avikchaudhuri reviewed (commented):
some questions

zou3519 commented on torchgen/model.py:
```diff
@@ -1901,6 +1901,8 @@ class BaseTy(Enum):
     SymInt = auto()
     SymBool = auto()
     GraphModule = auto()
+    NoneType = auto()
+    Operator = auto()
```
These are not acceptable schema types, what are you doing?

yiming0416 replied:
I'm trying to make `hop_schema_from_fx_node()` work for more types of hops, and I noticed some hops have either Arguments or Returns of type `None` or `Operator`

If this is not acceptable, is there a way that we can support schema generation for these types of hops?

ydwu4 replied:
If we want to fully roll out hop_schema_from_fx_node, we need to 1. add these types, and 2. support a tuple of mixed typed elements.

For Yiming: discussed with Richard today: we agree that the mixed dtype tuple type (and the None and operator types) are a large commitment. Other alternatives (e.g. spliting the lists into homogeneous types etc) are also tedious to implement. Using meta["val"] might be good enough for now If all we want is to determine the return type is single tensor or not. Let's wait for more use cases of schema and decide whether we want to commit.

yiming0416 replied:
Sounds good. Thanks for the clarification.

yiming0416 committed:
[export] Unify single and multiple return for hops (#143227)

Summary:


Introduce `is_hop_single_tensor_return` field to the `Node` class in serialization so that during deserialization when there is a single return, we know whether it is a tuple of a single element or a single element.

Test Plan:
```
buck2 run mode/dev-nosan sigmoid/inference/test:e2e_test_cpu -- -r E2ETestCPUCond
buck2 run mode/dev-nosan sigmoid/inference/test:test_passes -- -r test_const_folding2
```

Reviewed By: zhxchen17

Differential Revision: D66991624
